presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/KeyBasedSampler.java

```diff
@@ -0,0 +1,270 @@
+            if (!sampledFields.isEmpty()) {
+                warningCollector.add(new PrestoWarning(SAMPLED_FIELDS, String.format("Sampled the following columns/derived columns at %s percent:\n\t%s", SystemSessionProperties.getKeyBasedSamplingPercentage(session) * 100., String.join("\n\t", sampledFields))));
+            }
+            else {
+                warningCollector.add(new PrestoWarning(SEMANTIC_WARNING, "Sampling could not be performed due to the query structure"));
+            }
+
+            return rewritten;
+        }
+
+        return plan;
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final PlanNodeIdAllocator idAllocator;
+        private final List<String> sampledFields;
+
+        private Rewriter(Session session, FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator idAllocator, List<String> sampledFields)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.sampledFields = requireNonNull(sampledFields, "sampledFields is null");
+        }
+
+        private PlanNode addSamplingFilter(PlanNode tableScanNode, Optional<VariableReferenceExpression> rowExpressionOptional, FunctionAndTypeManager functionAndTypeManager)
+        {
+            if (!rowExpressionOptional.isPresent()) {
+                return tableScanNode;
+            }
+            RowExpression rowExpression = rowExpressionOptional.get();
+            RowExpression arg;
+            Type type = rowExpression.getType();
+            if (!Varchars.isVarcharType(type)) {
+                arg = call(
+                        "CAST",
+                        functionAndTypeManager.lookupCast(CAST, rowExpression.getType().getTypeSignature(), VARCHAR.getTypeSignature()),
+                        VARCHAR,
+                        rowExpression);
+            }
+            else {
+                arg = rowExpression;
+            }
+
+            RowExpression sampledArg = call(
+                    functionAndTypeManager,
+                    SystemSessionProperties.getKeyBasedSamplingFunction(session),
+                    DOUBLE,
+                    arg);
+
+            RowExpression predicate = call(
+                    LESS_THAN_OR_EQUAL.name(),
+                    functionAndTypeManager.resolveOperator(OperatorType.LESS_THAN_OR_EQUAL, fromTypes(DOUBLE, DOUBLE)),
+                    BOOLEAN,
+                    sampledArg,
+                    new ConstantExpression(SystemSessionProperties.getKeyBasedSamplingPercentage(session), DOUBLE));
```

Comment:
static import

Reply:
Done. Irritating that Intellij is inconsistent with this.